File: php/src/Amen/Webhooks/Signature.php
<?php
declare(strict_types=1);
namespace Amen\Webhooks;
/**
 * Signature verification for Amen webhook deliveries.
 *
 * Real Amen scheme (verified 2026-08-30):
 *   - Header  X-Webhook-Signature: sha256=<hex>   (HMAC-SHA256, hex, "sha256=" prefixed)
 *   - Header  X-Webhook-Timestamp: <ISO-8601>
 *   - Header  X-Webhook-Event:     <event type>   (also present as "event" in the JSON body)
 *   - signedMessage = timestamp . "." . rawBody
 *   - signature     = "sha256=" . hash_hmac('sha256', signedMessage, secret)
 */
final class Signature
{
    public const HEADER_SIGNATURE = 'X-Webhook-Signature';
    public const HEADER_TIMESTAMP = 'X-Webhook-Timestamp';
    public const HEADER_EVENT = 'X-Webhook-Event';
    public const ALGORITHM = 'sha256';

    /** Hex HMAC over `timestamp . "." . rawBody`. Returns the bare hex digest (no "sha256=" prefix). */
    public static function compute(string $secret, string $timestamp, string $rawBody, string $algorithm = self::ALGORITHM): string
    {
        return hash_hmac($algorithm, $timestamp . '.' . $rawBody, $secret);
    }

    /** Convenience: the digest with the "sha256=" prefix, exactly as sent in the X-Webhook-Signature header. */
    public static function sign(string $secret, string $timestamp, string $rawBody, string $algorithm = self::ALGORITHM): string
    {
        return $algorithm . '=' . self::compute($secret, $timestamp, $rawBody, $algorithm);
    }

    /**
     * Constant-time verification. Accepts hex digests, optionally prefixed like "sha256=<hex>".
     * Returns false when either the received signature or the timestamp is empty.
     */
    public static function verify(string $secret, string $timestamp, string $rawBody, ?string $received, string $algorithm = self::ALGORITHM): bool
    {
        if ($received === null || $received === '' || $timestamp === '') return false;
        $given = strtolower(trim(str_contains($received, '=') ? explode('=', $received, 2)[1] : $received));
        return hash_equals(self::compute($secret, $timestamp, $rawBody, $algorithm), $given);
    }
}

File: swift/Sources/AmenClient/Webhooks/Signature.swift
import Foundation
import CryptoKit

/// Signature verification over `timestamp + "." + rawBody`. Matches real Amen deliveries (docs/04-webhooks.md).
/// Delivery headers: `X-Webhook-Signature: sha256=<hex>`, `X-Webhook-Timestamp: <ISO-8601>`, `X-Webhook-Event: <type>`.
public enum WebhookSignature {
    public static let signatureHeader = "X-Webhook-Signature"
    public static let timestampHeader = "X-Webhook-Timestamp"
    public static let eventHeader = "X-Webhook-Event"

    /// Lowercase hex of HMAC-SHA256 over `utf8(timestamp) + "." + rawBody`.
    public static func compute(secret: String, timestamp: String, rawBody: Data) -> String {
        var signed = Data(timestamp.utf8)
        signed.append(0x2E) // "."
        signed.append(rawBody)
        return HMAC<SHA256>.authenticationCode(for: signed, using: SymmetricKey(data: Data(secret.utf8))).map { String(format: "%02x", $0) }.joined()
    }

    /// The full header value: `sha256=<hex>`.
    public static func signatureHeaderValue(secret: String, timestamp: String, rawBody: Data) -> String {
        "sha256=" + compute(secret: secret, timestamp: timestamp, rawBody: rawBody)
    }

    /// Accepts hex digests, optionally prefixed like "sha256=<hex>". Constant-time comparison.
    /// Returns false if either the received signature or the timestamp is nil/empty.
    public static func verify(secret: String, timestamp: String?, rawBody: Data, received: String?) -> Bool {
        guard let received, !received.isEmpty, let timestamp, !timestamp.isEmpty else { return false }
        let given = (received.hasPrefix("sha256=") ? String(received.dropFirst("sha256=".count)) : received).trimmingCharacters(in: .whitespaces).lowercased()
        let expected = compute(secret: secret, timestamp: timestamp, rawBody: rawBody)
        guard given.utf8.count == expected.utf8.count else { return false }
        return zip(given.utf8, expected.utf8).reduce(0) { $0 | ($1.0 ^ $1.1) } == 0
    }
}
